crates/saya/core/src/lib.rs

```diff
@@ -113,12 +114,35 @@ impl Saya {
                 continue;
             }
 
-            let fetched_block = self.provider.fetch_block(block).await?;
-
-            self.process_block(block, (&fetched_block, previous_block, genesis_state_hash)).await?;
+            // Fetch all blocks from the current block to the latest block
+            let fetched_blocks = future::try_join_all(
+                (block..latest_block).map(|block_number| self.provider.fetch_block(block_number)),
+            )
+            .await?;
+
+            // shift the state roots to the right by one, as proof of each block is based on the previous state root
+            let mut state_roots =
+                fetched_blocks.iter().map(|b| b.header.header.state_root).collect::<Vec<_>>();
+            state_roots.insert(0, previous_block_state_root);
+            previous_block_state_root = state_roots.pop().unwrap();
+
+            let params = fetched_blocks
+                .into_iter()
+                .zip(state_roots)
+                .map(|(b, s)| (b, s, genesis_state_hash))
+                .collect::<Vec<_>>();
+
+            let mut processed = Vec::with_capacity(params.len());
+            for p in params.clone() {
+                let prover_input = self.process_block(block, p).await?;
+                if let Some(input) = prover_input {
+                    processed.push(input);
+                }
+                block += 1;
```

Comment:
Here we don't take advantage of the parallelisation of the proof generation. Not too much a problem for now. But we should think about the code related to the comment made above. If we do a first version with Tokio tasks for instance, we rather want here a `wait` on the tasks to continue to compute the inner node (in the example with 2 leaves) once leaves proof is available.

Reply:
Below, I commented out the proving part of  `process_block`, so only the preparation of prover input is sequential here. I believe it should be due to the usage of `self.blockchain`, which would need to be shared and mutexed (so still sequential) otherwise.
The proof generation itself would be a part of the scheduler and work fully in parallel.